src/cpp/src/rag/text_embedding_pipeline.cpp

```diff
@@ -0,0 +1,304 @@
+#include "openvino/op/convert.hpp"
+#include "openvino/op/divide.hpp"
+#include "openvino/op/maximum.hpp"
+#include "openvino/op/multiply.hpp"
+#include "openvino/op/normalize_l2.hpp"
+#include "openvino/op/reduce_mean.hpp"
+#include "openvino/op/reduce_sum.hpp"
+#include "openvino/op/shape_of.hpp"
+#include "openvino/op/slice.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "utils.hpp"
+
+namespace {
+using namespace ov::genai;
+
+ov::AnyMap remove_config_properties(const ov::AnyMap& properties) {
+    auto properties_copy = properties;
+
+    properties_copy.erase(max_length.name());
+    properties_copy.erase(pooling_type.name());
+    properties_copy.erase(normalize.name());
+    properties_copy.erase(embed_instruction.name());
+    properties_copy.erase(query_instruction.name());
+
+    return properties_copy;
+}
+}  // namespace
+
+namespace ov {
+namespace genai {
+using utils::read_anymap_param;
+
+TextEmbeddingPipeline::Config::Config(const ov::AnyMap& properties) {
+    read_anymap_param(properties, ov::genai::max_length.name(), max_length);
+    read_anymap_param(properties, ov::genai::pooling_type.name(), pooling_type);
+    read_anymap_param(properties, ov::genai::normalize.name(), normalize);
+    read_anymap_param(properties, ov::genai::embed_instruction.name(), embed_instruction);
+    read_anymap_param(properties, ov::genai::query_instruction.name(), query_instruction);
+};
+
+class TextEmbeddingPipeline::TextEmbeddingPipelineImpl {
+public:
+    TextEmbeddingPipelineImpl(const std::filesystem::path& models_path,
+                              const std::string& device,
+                              const std::optional<Config>& config,
+                              const ov::AnyMap& properties = {})
+        : m_tokenizer{models_path} {
+        if (config.has_value()) {
+            m_config = *config;
+        }
+
+        ov::Core core = utils::singleton_core();
+
+        auto model = core.read_model(models_path / "openvino_model.xml", {}, properties);
+
+        model = apply_postprocessing(model, m_config);
+        if (m_config.max_length) {
+            m_tokenization_params.insert({ov::genai::max_length.name(), *m_config.max_length});
+        }
+
+        ov::CompiledModel compiled_model = core.compile_model(model, device, properties);
+
+        utils::print_compiled_model_properties(compiled_model, "text embedding model");
+        m_request = compiled_model.create_infer_request();
+    };
+
+    std::vector<EmbeddingResult> embed_documents(std::vector<std::string>& texts) {
+        start_embed_documents_async(texts);
+        return wait_embed_documents();
+    };
+
+    void start_embed_documents_async(std::vector<std::string>& texts) {
+        if (!m_config.embed_instruction) {
+            return start_embed_async(texts);
+        }
+
+        auto formatted_texts = format_texts(texts);
+        start_embed_async(formatted_texts);
+    };
```

Comment:
`format_texts()` already handles `!m_config.embed_instruction`. I think you can remove `!m_config.embed_instruction` part from `start_embed_documents_async()`

Reply:
It was a copy optimization attempt. I will apply it in further prs after testing if applicable.